pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -723,11 +721,77 @@ func isPodRunnableOrSucceeded(p *corev1.Pod) bool {
 	return p.Status.Phase != corev1.PodFailed
 }
 
+// Get the last timestamp on which the pod was observed active:
+// - the time the pod was declared Failed
+// - the deletion time
+func lastActiveTime(p *corev1.Pod) time.Time {
+	lastTransition := metav1.Now()
+	for _, c := range p.Status.Conditions {
+		if c.Type == corev1.ContainersReady {
+			if c.Status == corev1.ConditionFalse && c.Reason == string(corev1.PodFailed) {
+				lastTransition = c.LastTransitionTime
+			}
+			break
+		}
+	}
+	deletionTime := ptr.Deref(p.DeletionTimestamp, metav1.Now())
+	if lastTransition.Before(&deletionTime) {
+		return lastTransition.Time
+	}
+	return deletionTime.Time
+}
+
 // cleanupExcessPods will delete and finalize pods created last if the number of
 // activePods is greater than the totalCount value.
-func (p *Pod) cleanupExcessPods(ctx context.Context, c client.Client, r record.EventRecorder, totalCount int, activePods []corev1.Pod) error {
+func (p *Pod) cleanupExcessPods(ctx context.Context, c client.Client, r record.EventRecorder, totalCount int, activePods []corev1.Pod, inactivePods []corev1.Pod) error {
 	log := ctrl.LoggerFrom(ctx)
 
+	// Remove inactive pods that have a replacement
+	if len(inactivePods) > 0 && (len(inactivePods)+len(activePods)) > totalCount {
```

Comment:
I suppose here you are trying to identify which pod was replaced. I prefer we avoid having to maintain this code.

Reply:
Also refactored to reuse the deletion loop.